src/miniscript/mod.rs

```diff
@@ -44,48 +44,56 @@ use ToPublicKeyHash;
 use self::lex::{lex, TokenIter};
 use self::satisfy::{Satisfiable, Satisfier};
 use self::types::Property;
+use miniscript::types::Type;
+use miniscript::types::extra_props::ExtData;
 
 /// Top-level script AST type
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
-pub struct Miniscript<Pk, Pkh=hash160::Hash>(astelem::AstElem<Pk, Pkh>);
+pub struct Miniscript<Pk, Pkh=hash160::Hash>{
+    ///A node in the Abstract Syntax Tree(
+    pub node: astelem::AstElem<Pk, Pkh>,
+    ///The correctness and malleability type information for the AST node
+    pub ty: types::Type,
+    ///Additional information helpful for extra analysis.
+    pub ext: types::extra_props::ExtData,
+}
+
 
-impl<Pk, Pkh> From<astelem::AstElem<Pk, Pkh>> for Miniscript<Pk, Pkh>
+impl<Pk, Pkh> fmt::Debug for Miniscript<Pk, Pkh>
 where
     Pk: Clone + fmt::Debug,
     Pkh: Clone + fmt::Debug,
 {
-    fn from(t: astelem::AstElem<Pk, Pkh>) -> Miniscript<Pk, Pkh> {
-        let type_check = types::Type::type_check(&t, |_| None)
-            .expect("typecheck in Miniscript::from");
-        assert!(type_check.corr.base == types::Base::B);
-        Miniscript(t)
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self.node)
     }
 }
 
-impl<Pk, Pkh> fmt::Debug for Miniscript<Pk, Pkh>
-where
-    Pk: Clone + fmt::Debug,
-    Pkh: Clone + fmt::Debug,
+impl<Pk, Pkh> Miniscript<Pk, Pkh>
+    where Pk : fmt::Debug + fmt::Display + Clone, Pkh: fmt::Debug + fmt::Display + Clone
 {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{:?}", self.0)
+    /// Add type information(Type and Extdata) to Miniscript based on
+    /// `AstElem` fragment. Dependant on display and clone because it is used
+    /// in creating Miniscript from String
+    pub fn from_ast(t: astelem::AstElem<Pk, Pkh>) -> Result< Miniscript<Pk, Pkh>, Error> {
```

Comment:
I wonder if this should be `pub` or not, and whether it ought to return a `Box<Miniscript>` rather than a `Miniscript`.

Reply:
This is useful in compiler code. The comment to this code was incorrect. The `Debug` and `Display` bounds are because of the `Error` type `type_check` code.